### PR TITLE
[ML] Incremental train (part 1)

### PR DESCRIPTION
**Pull request by tveasey**
This makes some mainly interface changes in anticipation of incremental training. The plan is to adjust the loss function to include penalties for modifying the tree topology and results. It should be possible to manage this in `CBoostedTreeLeafNodeStatistics` and reuse our existing training code. As such I've made this a hierarchy. The current implementation moves into `CBoostedTreeLeafNodeStatisticsScratch` and I'll have a new implementation `CBoostedTreeLeafNodeStatisticsIncremental` (placeholder). There will likely need to be further tweaks down the line particularly to the workspace object, but I'll make those in the context of working on the implementation `CBoostedTreeLeafNodeStatisticsIncremental`. This feels like a useful place to break up the change before I move into implementation details.

A step towards #1721.

### CODE REVIEW

File: lib/api/CDataFrameTrainBoostedTreeRunner.cc [hunk 1]
```diff
@@ -358,7 +358,8 @@ std::size_t CDataFrameTrainBoostedTreeRunner::estimateBookkeepingMemoryUsage(
     std::size_t totalNumberRows,
     std::size_t /*partitionNumberRows*/,
     std::size_t numberColumns) const {
-    return m_BoostedTreeFactory->estimateMemoryUsage(
+    // TODO Incremental training.
```
Comment (valeriy42) [inline lib/api/CDataFrameTrainBoostedTreeRunner.cc:361]: Do you want to reference the issue, so this TODO won't get lost?

**Review by valeriy42 — APPROVED** [review]
LGTM

File: lib/api/CDataFrameTrainBoostedTreeRunner.cc (hunk 1 above)
Comment (tveasey) [reply lib/api/CDataFrameTrainBoostedTreeRunner.cc]: [f55e90f](https://github.com/elastic/ml-cpp/pull/1772/commits/f55e90f939a5e446e72435c87fd0a38007525164)

**Comment by tveasey** [comment]
Windows CI failure is known worker issue.
